roxie/ccd/ccdkey.cpp

```diff
@@ -1593,393 +1065,134 @@ class InMemoryIndexCursor : implements IInMemoryIndexCursor, public CInterface
-    {
-        unsigned lim = segMonitors.length();
-        while (segno < lim)
-            segMonitors.item(segno++).setLow(keyBuffer);
-    }
-
-    void endRange(unsigned segno) const
+    ~InMemoryIndexCursor()
     {
-        unsigned lim = segMonitors.length();
-        while (segno < lim)
-            segMonitors.item(segno++).endRange(keyBuffer);
     }
 
-    bool incrementKey(unsigned segno, const void *rowval) const
+    // interface IDirectReader
+    virtual bool isKeyed() const override
     {
-        // Increment the key buffer to next acceptable value after rowval
-        for(;;)
-        {
-            IKeySegmentMonitor &seg = segMonitors.item(segno);
-            if (rowval)
-                seg.copy(keyBuffer, rowval);
-            if (seg.increment(keyBuffer))
-            {
-                setLow(segno+1);
-                return true;
-            }
-            if (!segno)
-                return false;
-            segno--;
-        }
+        // Could argue should return false if no indexed fields. We still use a key as it avoids prefetches for variable-size rows
+        // The value is used both to decide if worth doing in parallel (where false would be more appropriate)
+        // and for continuation serialization (where false would generate invalid results)
+        return true;
     }
 
-    bool matches(const RtlRow *r, const SegMonitorArray &segs) const
+    virtual IDirectStreamReader *queryDirectStreamReader()
     {
-        ForEachItemIn(idx, segs)
-        {
-            if (!segs.item(idx).matches(r))
-                return false;
-        }
-        return true;
+        throwUnexpected();
     }
 
-    unsigned firstInRange(unsigned a) const
+    virtual const byte *nextRow() override
     {
-        if (!segMonitors.length())
-            return a;
-        for (;;)
+        while (!eof)
         {
-            // binchop...
-            int b = numPtrs;
-            int rc;
-            while ((int)a<b)
+            if (!keySearcher->next())
             {
-                int i = a+(b-a)/2;
-                rc = docompare(keyBuffer, GETROW(i));
-                if (rc>0)
-                    a = i+1;
-                else
-                    b = i;
+                eof = true;
+                break;
             }
-            if (a<numPtrs)
+            else if (postFilter.matches(keySearcher->queryRow()))
             {
-                const void *row = GETROW(a);
-                unsigned seg = 0;
-                unsigned lim = segMonitors.length();
-                for (;;)
+                if (translator)
                 {
-                    if (!segMonitors.item(seg).matchesBuffer(row))  // Technically should be using RtlRow here but we don't support varoffset in indexes yet
-                        break;
-                    seg++;
-                    if (seg==lim)
-                        return a;
+                    MemoryBufferBuilder aBuilder(buf, 0);
```

Comment:
I think buf needs to have length set back to 0 in finishedRow()

Reply:
Probably